File: 0001 Two Sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Using Map
        // Time: O(n)
        // Space: O(n)
        Map<Integer, Integer> d = new HashMap<Integer, Integer>();
        int[] ret = new int[2];

        for (int i=0; i<nums.length; i++) {
            int rem = target - nums[i];

            if (d.get(rem) != null) {
                ret[0] = i;
                ret[1] = d.get(rem);
                return ret;
            }

            d.put(nums[i], i);
        }

        return ret;

    }
}

File: 0001 Two Sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Using map
        # Time: O(n)
        # Space: O(n)
        map = {}
        n = len(nums)

        for index in range(n):
            rem = target - nums[index]
            if rem in map:
                return [map[rem], index]
            map[nums[index]] = index

        return []
        
        '''
        # Using Counter and index
        # Time: O(n^3)
        # Space: O(1)
        for i in range(0, len(nums)):
            if (nums.count(target - nums[i]) > 0 and i != nums.index(target - nums[i])):
                return [i, nums.index(target - nums[i])]
        '''
